app-backend/common/src/main/scala/ast/package.scala

```diff
@@ -0,0 +1,73 @@
+package com.azavea.rf.common
+
+import com.azavea.rf.database.Database
+import com.azavea.rf.database.tables.{ToolRuns, Tools}
+import com.azavea.rf.datamodel.{Tool, ToolRun, User}
+import com.azavea.rf.tool.ast.MapAlgebraAST
+import com.azavea.rf.tool.ast.MapAlgebraAST._
+import com.azavea.rf.tool.eval.{ASTDecodeError, DatabaseError, Interpreter}
+import com.azavea.rf.tool.params.EvalParams
+
+import cats._
+import cats.data._
+import cats.implicits._
+import cats.data.Validated.{Invalid, Valid}
+import io.circe._
+
+import java.lang.IllegalArgumentException
+import java.util.UUID
+import scala.concurrent.{ExecutionContext, Future}
+
+package object ast {
+
+  /** Collect resources necessary to carry out substitutions on  ASTs */
+  def assembleSubstitutions(
+    ast: MapAlgebraAST,
+    user: User,
+    assembled: Map[UUID, MapAlgebraAST] = Map()
+  )(implicit database: Database, ec: ExecutionContext): OptionT[Future, Map[UUID, MapAlgebraAST]] =
+    ast match {
+      case ToolReference(id, refId) =>
+        OptionT(Tools.getTool(refId, user)).map({ referent =>
+          val astPatch = referent.definition.as[MapAlgebraAST].valueOr(throw _)
+          if (assembled.keys.toList.contains(refId))
+            throw new IllegalArgumentException(s"Repeated substitution $refId found; likely cyclical reference")
```

Comment:
This is not the ideal way to solve this problem - I'd like to see it solved with more consistent and complete usage of `Invalid` or `Either` eventually